### PR TITLE
`General`: Add dynamic tab titles

### PR DESCRIPTION
**Pull request by mallgaier**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context
Currently, Artemis only provides static tab titles defined in the router modules. 
The static tab titles make navigation difficult when several tabs are open. If several tabs with the same title (e.g. "Programming Exercise") are open at the same time, navigation between the tabs is very limited.

### Description
This PR extends the breadcrumb implementation of the client. Despite the most specific title of the respective page (either dynamically (e.g. exercise title) or statically) , the exercise title, the exam, the lecture title and the course title are (optionally) included. The tab titles are build from the most specific to the most general title, separate by a `|`. 
As a fallback, the statically defined titles in the router modules are used. 

### Steps for Testing
Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Navigate around and make sure, that the tab titles have a meaningful and hierarchical structure.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2
#### Exam Mode Test
- [x] Test 1
- [ ] Test 2

### Screenshots

image
image
image
image

## Summary by CodeRabbit

- **New Features**
	- Enhanced the navigation bar to dynamically display titles for courses, exercises, lectures, and exams, improving user navigation and context awareness.

### CODE REVIEW

File: src/main/webapp/app/shared/layouts/navbar/navbar.component.ts [hunk 1]
```diff
@@ -831,6 +841,59 @@ export class NavbarComponent implements OnInit, OnDestroy {
...
+        // Include the most specific title into the tab title, but only if the title is meant to be displayed to the user, i.e. should be translated.
+        const generalTitle = this.breadcrumbs[this.breadcrumbs.length - 1].translate
+            ? this.translateService.instant(this.breadcrumbs[this.breadcrumbs.length - 1].label)
+            : undefined;
+        const titles = [generalTitle, this.exerciseTitle, this.examTitle, this.lectureTitle, this.courseTitle].filter((title) => title !== undefined).join(' | ');
+        // No need have a dynamic title on the start page -> use the title defined in the Router modules.
+        if (titles && this.breadcrumbs.length > 1) {
+            this.titleService.setTitle(titles);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mallgaier) [reply src/main/webapp/app/shared/layouts/navbar/navbar.component.ts:860]: Not needed, as the default titles are defined in the router modules.

**Review by AntonGeTUM — COMMENTED** [review]
Tested on ts2, works fine for the most part, here some things I noticed:
- for programming exercises: only 'Edit in editor' doesn't have a tab title
- individual competencies don't have tab titles
- when creating exercises or lectures, the title is just 'Create'

**Comment by egekurt123** [comment]
Tested on ts3; besides Anton´s comments, Create Exam Page also just has "Create" as a tab title. Apart from that, it looks great.

**Review by pzdr7 — COMMENTED** [review]
Tested on TS3. The exercise import functionality is not entirely consistent:

Navigate to the management view of a course > Exercises

- Import Programming -> Title: `Import | (name of the course)` 
- Import Quiz -> Title: `(name of the course)`
- Import Modeling, File Upload, Text exercise -> Title: `(name of the imported exercise) | (name of the course)`

Maybe it could be `Import (type of the exercise) | (name of the course)`?

(note: In exams > exercise groups, it's `Import | (title of the exam) | (name of the course)` regardless of the exercise type)

**Review by pzdr7 — APPROVED** [review]
Tested on TS3. All in all, looks good. We can work out any remaining inconsistencies in follow-up PRs 👍

**Pull request merged**